doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,205 @@
+
+### Turning off GHC's pre-inliner
+
+As said before, one reason why the semantics of Plutus Tx is difficult to predict is because we run GHC's pre-inliner, which may perform unconditionally inlining.
+Since GHC 9.0, GHC provides [an option](https://hackage.haskell.org/package/ghc-9.6.1/docs/GHC-Core-Opt-Simplify-Env.html#v:sm_pre_inline) to turn it off when running the simplifier.
+By turning it off, GHC (hopefully) won't inline anything at all, which means it won't change the strictness of Plutus Tx programs.
+
+At this time we can't simply turn the pre-inliner off, because doing so breaks the Plutus Tx compiler.
+For details, see [this note](https://github.com/input-output-hk/plutus/blob/dc57fe1b8497835ef2a7794c017c5a50c0b7e647/plutus-tx-plugin/src/PlutusTx/Plugin.hs#L123).
+The problem described in the note should, however, be resolved in a different way, as explained in the note.
+Once done, we can go ahead and turn it off.
+
+### Performing GHC-like Inlining for Functions with the INLINE Pragma
+
+In Plutus Tx, like most other strict languages, only a few built-in constructs are non-strict (such as if-then-else).
+It is cumbersome, if not impossible, to create user-defined non-strict functions.
+The `nonstrict` function described before does not help because it cannot be used in defining a function.
+It would be useful to allow user-defined non-strict functions, so that we can define functions like `&&` and `||` that short-circuit properly.
+Some languages like Scala have special syntaxes for defining non-strict functions; we can't change Haskell's syntax, so we need to find a different solution.
+
+The idea is to make the Plutus Tx compiler perform GHC-like inlining for functions with the INLINE pragma.
+
+"GHC-like inlining", or "inlining in the GHC way", for lack of a better term, refers to inlining in a way similar to how GHC does it, i.e., replacing a variable with its RHS, then performing beta substitution.
+As an example, consider the definition of `(&&)`:
+
+```haskell
+(&&) :: Bool -> Bool -> Bool
+a && b = if a == False then False else b
+{-# INLINE && #-}
+```
+
+When the PIR inliner inlines `(&&)` in the normal way, `expr1 && expr2` becomes
+
+```haskell
+let !x = expr1
+    !y = expr2
+ in if x == False then False else y
+```
+
+whereas inlining `(&&)` in the GHC way yields
+
+```haskell
+if expr1 == False then False else expr2
+```
+
+Normally, it is incorrect for the Plutus Tx compiler to inline `(&&)` in the GHC way, because in the `expr1 && expr2` example, before inlining, it evaluates `expr2` strictly, but after GHC-like inlining, `expr2` is evaluated non-strictly.
+But because of the `INLINE` pragma, the Plutus Tx compiler should inline `(&&)` in the GHC way.
+And this is desirable because we want `(&&)` to be non-strict.
+Functions that should not be inlined in the GHC way should use the `INLINABLE` pragma.
```

Comment:
Why shouldn't INLINE mean exactly what it means today? i.e. I'm not sure why this alternative meaning of inlining should get to claim the pragma, indeed I think that's surprising. We could use an alternative GHC annotation, perhaps?

Reply:
@michaelpj  you mean this https://downloads.haskell.org/ghc/latest/docs/users_guide/extending_ghc.html?highlight=ann%20pragma#source-annotations  ? That's a nice idea, and better than giving special meanings to INLINE/INLINEABLE pragmas

Reply:
> Why shouldn't INLINE mean exactly what it means today?

INLINE should mean exactly what _GHC_ means by INLINE. Which is different from what our inliner do. See the example above. We want the second result, not the first.

Reply:
hm....it does seem a bit confusing. So, if users want to define non-strict functions, they use the `INLINE` pragma. Otherwise, they use the `INLINEABLE` pragma, and it will be inlined with our PIR inliner?

Reply:
That's not what GHC means by `INLINE`. GHC means the same thing that we do. The fact that GHC's simplifier then inlines things is just a weird byproduct. `INLINE` should mean "inline this", nothing else.

Reply:
And yes, I think source annotations could be a good tool here.